pkg/controllers/util/federatedclient/podinformer_test.go

```diff
@@ -0,0 +1,346 @@
+			clusters:            4,
+			podNum:              3000,
+			expectCacheSynced:   true,
+			availablePodListers: 0,
+			enableSimplifying:   true,
+		},
+		{
+			name:                "4 clusters, regular pod, limit 2",
+			clusters:            4,
+			podNum:              3000,
+			expectCacheSynced:   true,
+			availablePodListers: 2,
+			enableSimplifying:   false,
+		},
+		{
+			name:                "4 clusters, regular pod, no limit",
+			clusters:            4,
+			podNum:              3000,
+			expectCacheSynced:   true,
+			availablePodListers: 0,
+			enableSimplifying:   false,
+		},
+	}
+
+	type memberCluster struct {
+		name              string
+		client            kubeclient.Interface
+		informer          kubeinformer.SharedInformerFactory
+		enableSimplifying bool
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			defer func() {
+				if e := recover(); e != nil {
+					if err, ok := e.(error); ok && errors.Is(err, errLimitExceeded) {
+						t.Errorf("failed to rate limit for listing pod")
+						return
+					}
+					t.Errorf("unexpected panic: %v", e)
+				}
+			}()
+			var sem *semaphore.Weighted
+			if tt.availablePodListers > 0 {
+				sem = semaphore.NewWeighted(tt.availablePodListers)
+			}
+			ctx := context.Background()
+			clusters := make([]memberCluster, tt.clusters)
+			for i := 0; i < tt.clusters; i++ {
+				var client kubeclient.Interface
+				if tt.availablePodListers == 0 {
+					_, client = setupFakeClient(int64(tt.clusters), tt.podNum)
+				} else {
+					_, client = setupFakeClient(tt.availablePodListers, tt.podNum)
+				}
+				informer := kubeinformer.NewSharedInformerFactory(client, util.NoResyncPeriod)
+				clusters[i] = memberCluster{
+					name:              fmt.Sprintf("member-cluster-%d", i),
+					client:            client,
+					informer:          informer,
+					enableSimplifying: tt.enableSimplifying,
+				}
+			}
+
+			for i := range clusters {
+				addPodInformer(ctx, clusters[i].informer, clusters[i].client, sem, clusters[i].enableSimplifying)
+				clusters[i].informer.Start(ctx.Done())
+			}
+			var wg sync.WaitGroup
+			wg.Add(tt.clusters)
+			for i := range clusters {
+				go func(i int) {
+					defer wg.Done()
+
+					podLister := clusters[i].informer.Core().V1().Pods().Lister()
+					podsSynced := clusters[i].informer.Core().V1().Pods().Informer().HasSynced
+					ctx, cancel := context.WithTimeout(ctx, 5*time.Second)
+					defer cancel()
+
+					// test lister
+					if !cache.WaitForNamedCacheSync(clusters[i].name, ctx.Done(), podsSynced) {
+						if tt.expectCacheSynced {
+							t.Errorf("%s should be synced, but it was not\n", clusters[i].name)
+						}
+						return
+					}
+					if !tt.expectCacheSynced {
+						t.Errorf("%s should not be synced, but it was\n", clusters[i].name)
+						return
+					}
+					pods, err := podLister.List(labels.Everything())
+					if err != nil {
+						t.Errorf("%s: unexpected error when listing pod: %v\n", clusters[i].name, err)
```

Comment:
```suggestion
						t.Fatalf("%s: unexpected error when listing pod: %v\n", clusters[i].name, err)
```
Generally when an error occurs, there's little reason to continue with the rest of the test case.

Reply:
Preferably all `Errorf` should be replaced with `Fatalf`